netty/src/test/java/io/grpc/netty/NettyClientTransportTest.java

```diff
@@ -538,6 +550,88 @@ public void maxHeaderListSizeShouldBeEnforcedOnClient() throws Exception {
     }
   }
 
+  @Test
+  public void huffmanCodingShouldBePerformed() throws Exception {
+    startServer();
+
+    Callable<NettyClientTransport> huffmanTransport =  () -> newTransport(newNegotiator(),
+        DEFAULT_MAX_MESSAGE_SIZE, DEFAULT_MAX_HEADER_LIST_SIZE, 1, null, true,
+        TimeUnit.SECONDS.toNanos(10L), TimeUnit.SECONDS.toNanos(1L),
+        new ReflectiveChannelFactory<>(NioSocketChannel.class), group);
+
+    Metadata aHeaders = new Metadata();
+    aHeaders.put(Metadata.Key.of("test", Metadata.ASCII_STRING_MARSHALLER),
+        LONG_STRING_OF_A);
+
+    Metadata tildeHeaders = new Metadata();
+    tildeHeaders.put(Metadata.Key.of("test", Metadata.ASCII_STRING_MARSHALLER),
+                  LONG_STRING_OF_TILDE);
+
+    Metadata warmupHeaders = new Metadata();
+    warmupHeaders.put(Metadata.Key.of("test", Metadata.ASCII_STRING_MARSHALLER), "unused");
+
+    getRpcSize(huffmanTransport, warmupHeaders);
+
+    long aHeaderRpcSize = getRpcSize(huffmanTransport, aHeaders);
+
+    long tildeHeaderRpcSize = getRpcSize(huffmanTransport, tildeHeaders);
+
+    assertThat(aHeaderRpcSize).isLessThan(tildeHeaderRpcSize);
+  }
+
+  @Test
+  public void huffmanCodingShouldNotBePerformed() throws Exception {
+    startServer();
+
+    Callable<NettyClientTransport> nonHuffmanTransport =  () -> newTransport(newNegotiator(),
+        DEFAULT_MAX_MESSAGE_SIZE, DEFAULT_MAX_HEADER_LIST_SIZE, Integer.MAX_VALUE, null, true,
+        TimeUnit.SECONDS.toNanos(10L), TimeUnit.SECONDS.toNanos(1L),
+        new ReflectiveChannelFactory<>(NioSocketChannel.class), group);
+
+    Metadata aHeaders = new Metadata();
+    aHeaders.put(Metadata.Key.of("test", Metadata.ASCII_STRING_MARSHALLER),
+        LONG_STRING_OF_A);
+
+    Metadata tildeHeaders = new Metadata();
+    tildeHeaders.put(Metadata.Key.of("test", Metadata.ASCII_STRING_MARSHALLER),
+        LONG_STRING_OF_TILDE);
+
+    Metadata warmupHeaders = new Metadata();
+    warmupHeaders.put(Metadata.Key.of("test", Metadata.ASCII_STRING_MARSHALLER), "unused");
+
+    //warm up hpack
+    getRpcSize(nonHuffmanTransport, warmupHeaders);
+
+    long aHeaderRpcSize = getRpcSize(nonHuffmanTransport, aHeaders);
+    long tildeHeaderRpcSize = getRpcSize(nonHuffmanTransport, tildeHeaders);
+
+    assertThat(aHeaderRpcSize).isEqualTo(tildeHeaderRpcSize);
+  }
+
+  @CanIgnoreReturnValue
+  private long getRpcSize(Callable<NettyClientTransport> transportSupplier, Metadata headers)
+      throws Exception {
+
+    NettyClientTransport transport = transportSupplier.call();
+
+    callMeMaybe(transport.start(clientTransportListener));
+
+    ChannelTrafficShapingHandler channelTrafficShapingHandler =
+        new ChannelTrafficShapingHandler(0);
+
+    transport.channel().pipeline().addFirst(channelTrafficShapingHandler);
+
+    new Rpc(transport, headers).halfClose().waitForResponse();
+
+    Uninterruptibles.sleepUninterruptibly(1, TimeUnit.MILLISECONDS);
```

Comment:
There appears to be a race between the RPC completing/releasing this thread and the traffic counter finishing the counting of the bytes. I'm working to remove this hack.